ProcessHacker/memprv.c

```diff
@@ -239,6 +239,65 @@ PPH_MEMORY_ITEM PhpSetMemoryRegionType(
     return memoryItem;
 }
 
+void PhpUpdateHeapRegions(
+    _In_ PPH_MEMORY_ITEM_LIST List,
+    _In_ HANDLE ProcessHandle
+    )
+{
+    PRTL_DEBUG_INFORMATION debugBuffer;
+    HANDLE powerRequestHandle = NULL;
+
+    if (WindowsVersion >= WINDOWS_8)
+    {
+        // The RtlQueryProcessDebugInformation can cause deadlock when querying frozen (suspended) immersive process.
+        // A workaround was implemented using PhCreateExecutionRequiredRequest().
+        PhCreateExecutionRequiredRequest(ProcessHandle, &powerRequestHandle);
+    }
+
+    if (debugBuffer = RtlCreateQueryDebugBuffer(0x10000, FALSE))
+    {
+        if (NT_SUCCESS(RtlQueryProcessDebugInformation(List->ProcessId, RTL_QUERY_PROCESS_HEAP_SUMMARY | RTL_QUERY_PROCESS_HEAP_ENTRIES_EX, debugBuffer)))
```

Comment:
`RTL_QUERY_PROCESS_HEAP_ENTRIES_EX` flag (0x200) does in fact return just information about heap segments (instead of all busy/free allocations).
Thus maybe `RTL_QUERY_PROCESS_HEAP_SEGMENTS` would be more appropriate name?

Reply:
Don't hard-code the commit size with 0x10000... It's required to loop and increment just like kernel32 does with Heap32First/Heap32Next: 

https://github.com/processhacker/processhacker/blob/b228cb0d2aa8035ae57a765c555bb1cf70be1de5/phlib/native.c#L10267-L10270

> HEAP_SEGMENTS would be more appropriate name

Yeah we can rename it.

Reply:
> Don't hard-code the commit size with 0x10000... It's required to loop and increment just like kernel32 does with Heap32First/Heap32Next:

Actually I wanted to make the code simpler here. Since we are interested only in heap segments, the returned debug buffer is usually quite small even for processes that use heap heavily (it contains just few records for each heap).

So it will practically always fit the 64k reserved debug buffer (which also seems to be smallest value that works with `RtlQueryProcessDebugInformation`).

Of course I can copy the code from `PhQueryProcessHeapInformation` if you think it would be better.

Though in that case it may be better to extract it to some helper function so that it can be use in both `PhQueryProcessHeapInformation` and `PhpUpdateHeapRegions`.
Not sure what would be the best place for such _helper_.

Reply:
@dmex 
> Yeah we can rename it.

Will you do that?
Or should I do it (within this PR)?

Reply:
> So it will practically always fit the 64k reserved debug buffer

I had to debug this function and flags last week because of insider build changes and fixed commit sizes just doesn't work with some processes. The size loop was directly copied from Heap32First/Heap32Next in kernel32.dll, so the correct usage of this function requires the loop.

> Or should I do it (within this PR)?

 I'll update the flags separately because there are other changes required for insider builds:

![image](https://user-images.githubusercontent.com/1306177/147667422-bd706dfb-76c4-4564-855f-1be592a398a1.png)